[EASI-3360] GraphQL schema for TRB Admin Note category-specific data

DylanSpOddball opened the pull request:
# EASI-3360

## Changes and Description

- Update the GraphQL schema to include category-specific data on TRB admin notes (i.e. notes in the Initial Request Form category can select which sections the note applies to)
  - For queries:
    - Each category has a type for the data that's specific to that category (including the General Request and Consult Session categories, which _don't_ have any category-specific data).
    - The `TRBAdminNote` type has a `categorySpecificData` field whose type is a [GraphQL union](https://graphql.org/learn/schema/#union-types) of the category-specific types.
  - For mutations:
    - Each category has its own mutation and associated input type, with the types having non-nullable fields for required data.
      - The resolvers for the new mutations have placeholder implementations that call the existing `CreateTRBAdminNote()` function, so the frontend can more easily code against them in EASI-3361.
    - The existing `createTRBAdminNote` mutation is kept around to avoid breaking existing code; it will be removed in EASI-3458.
    - The `updateTRBAdminNote` mutation and resolver don't currently account for category-specific data. That mutation isn't currently being used, but might be in the future (see [this Slack conversation](https://cmsgov.slack.com/archives/CNU2B59UH/p1697562260732339)); if we need to update it to allow editing category-specific data, we can change the schema at that time.
- Provide mock data for the `categorySpecificData` field's resolver on the backend, including category-specific data, for the frontend to use as an example to code against.
- Update the Postman collection to include an example of querying the union on the `TRBAdminNote` type, as well as saving requests to call `createTRBAdminNote` and `updateTRBAdminNote`.

## How to test this change

There's not much to test, since the bulk of this PR is GraphQL schema changes that haven't been implemented. However, we can test that the existing `createTRBAdminNote` calls still work, and that the placeholder data is returned.
1. Spin up the app, including the frontend.
2. Run `scripts/dev db:seed`.
3. In the UI, go to the TRB admin view, open "Case 7 - Advice letter in review".
4. View the advice letter and add a note (of any category). The note should be created without any issues, and should be able to be viewed without errors.
5. Open Postman, set the `trbRequestID` collection variable's current value to the ID of the "Case 7" request, and save that value.
6. Run the "Get Admin Notes, with category-specific fields" request (under TRB -> TRB Admin Notes). This should return the placeholder data from `schema.resolvers.go` for the note's category.

## Questions
1. As mentioned above, I'm not sure whether to have types for all five categories, including the categories that don't currently have category-specific data, and include them in the `TRBAdminNoteCategorySpecificData` union. Including all five makes it clear that it covers all categories, but having the extra types that don't have any useful data is annoying, especially because gqlgen doesn't support empty types (hence the `placeholderField` fields). It doesn't add any major burden either way to the backend; @adamodd, @aterstriep, @patrickseguraoddball, what do you think?
2. I've nested common fields used by all categories of advice notes in the `CreateTRBAdminNoteCommonFields` input type to avoid repetition and redundancy - @aterstriep, @adamodd, @patrickseguraoddball, does this sort of nesting work fine with frontend code that calls the mutation from a form?

## Notes
* Backend implementation notes (for EASI-3362):
  * We'll need to add fields to the `TRBAdminNote` Go model and SQL table for the data that can go right on the intake (such as the various "appliesTo*" checkboxes/booleans). The relations with TRB documents and recommendations will need to be handled by many-to-many link tables in SQL, with small Go models representing each link, similar to [`/pkg/models/trb_request_lcid.go`](https://github.com/CMSgov/easi-app/blob/main/pkg/models/trb_request_lcid.go), which links TRB requests and system intakes.
  * The SQL columns added to `trb_admin_notes` will be nullable, by necessity; we'll see in EASI-3362 what constraints can be enforced in SQL.
  * Existing admin notes can just have all `appliesTo*` checkboxes set to false (when applicable), with no links to documents/recommendations created for them.
  * More details are in [this Jira comment](https://jiraent.cms.gov/browse/EASI-3360?focusedId=2356345&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-2356345) on EASI-3360.

## Decisions after PR was initially posted for review
- Dev meeting on 10/17/23 decided to use multiple mutations, one per category, for creating admin notes.
  -  GraphQL doesn't currently support union types in input fields - the original approach was to have nullable fields covering all possible categories in `CreateTRBAdminNoteInput`, with a single resolver. We went with the approach of multiple mutations for greater type-safety.
- Clay and I decided on 10/17/23 to wait on implementing category-specific functionality for `updateTRBAdminNote` for now, after talking with UX.

DylanSpOddball commented:
Added @aterstriep since she paired with me on working out this GQL schema, added @ClayBenson94 for review of the overall design.

aterstriep commented:
Schema changes look good from a frontend perspective!

mynar7 commented on pkg/graph/schema.graphql:
```diff
@@ -2344,22 +2394,78 @@ type TRBAdviceLetter {
   trbRequestId: UUID!
   category: TRBAdminNoteCategory!
   noteText: HTML!
+
+  # General Request category - no additional data
+
+  # Initial Request Form category
+  initialRequestFormData: TRBAdminNoteInitialRequestFormInput
+
+  # Supporting Documents category
+  supportingDocumentsData: TRBAdminNoteSupportingDocumentsInput
+
+  # Consult Session category - no additional data
+
+  # Advice Letter category
+  adviceLetterData: TRBAdminNoteAdviceLetterInput
```
I think passing in nested objects for different form types introduces some issues. For anyone lacking the context, they might see this type schema and get confused on what is actually required for an admin note (without comments in FE and BE code) which undermines the benefit of the type system.

This also causes us to write validation code to ensure the correct info is being passed in per category that might not be obvious to the FE.

DylanSpOddball replied:
Definitely will add some comments here to clarify.

DylanSpOddball replied:
Updated in https://github.com/CMSgov/easi-app/pull/2231/commits/95bbf51eada4ee35909fd964d0c41ed5badb2013 - @mynar7, how does this look?

mynar7 replied:
Looks right so far!

mynar7 commented on pkg/graph/schema.graphql:
```diff
@@ -2306,6 +2306,55 @@ enum TRBAdminNoteCategory {
+  recommendations: [TRBAdviceLetterRecommendation!]!
+}
+
+union TRBAdminNoteCategorySpecificData = TRBAdminNoteGeneralRequestCategoryData | TRBAdminNoteInitialRequestFormCategoryData | TRBAdminNoteSupportingDocumentsCategoryData | TRBAdminNoteConsultSessionCategoryData | TRBAdminNoteAdviceLetterCategoryData
```
Would it make sense for category specific data to optionally be null instead of having a placeholder type? The category also lives on the parent note type which should let the FE know not to expect categorical data in those generic note cases, right?

DylanSpOddball replied:
It might; I was hoping to use the approach of making `TRBAdminNote` a union type to get around this, but that didn't work, per [this comment](https://github.com/CMSgov/easi-app/pull/2231#issuecomment-1765408993). The backend code would be pretty similar either way. I'm not sure what would be more ergonomic for the frontend code.

Example FE code for working with query results, with the union type having cases/variants for all five categories (the current schema), showing exhaustive checking for all possible cases:
```ts
if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteGeneralRequestCategoryData') {
  // do nothing, no actual data
} else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteInitialRequestFormCategoryData') {
  // Typescript now knows that adminNote.categorySpecificData has fields appliesToBasicRequestDetails, appliesToSubjectAreas, and appliesToAttendees
  // do something with those fields
} else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteSupportingDocumentsCategoryData' {
  // Typescript now knows that adminNote.categorySpecificData has a documents field
  // do something with those fields
} else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteConsultSessionCategoryData') {
  // do nothing, no actual data
} else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteAdviceLetterCategoryData') {
  // Typescript now knows that adminNote.categorySpecificData has fields appliesToMeetingSummary, appliesToNextSteps, recommendations
  // do something with those fields
}
```

Alternatively, if the GraphQL schema doesn't include the cases with no additional data, makes `adminNote.categorySpecificData` nullable, and defines `union TRBAdminNoteCategorySpecificData = TRBAdminNoteInitialRequestFormCategoryData | TRBAdminNoteSupportingDocumentsCategoryData | TRBAdminNoteAdviceLetterCategoryData`, the code might look like:
```ts
if (adminNote.categorySpecificData) {
  if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteInitialRequestFormCategoryData') {
    // Typescript now knows that adminNote.categorySpecificData has fields appliesToBasicRequestDetails, appliesToSubjectAreas, and appliesToAttendees
    // do something with those fields
  } else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteSupportingDocumentsCategoryData' {
    // Typescript now knows that adminNote.categorySpecificData has a documents field
    // do something with those fields
  } else if (adminNote.categorySpecificData.__typename === 'TRBAdminNoteAdviceLetterCategoryData') {
    // Typescript now knows that adminNote.categorySpecificData has fields appliesToMeetingSummary, appliesToNextSteps, recommendations
    // do something with those fields
  }
}
```

This avoids the messiness of needing placeholder fields, but the FE code doesn't necessarily clearly show that it's covering all categories. @aterstriep, @adamodd, what do you think?

adamodd replied:
Nullable `adminNote.categorySpecificData` looks neat to me, but either was is probably fine. Would `categorySpecificData` only need to show properties of one category data type? If so it's probably ok that FE can't see them all.

DylanSpOddball replied:
@adamodd `categorySpecificData` would only show properties of one category's data at a time. Before checking `__typename`, no fields on it would be accessible; once the value of `__typename` was checked, Typescript would allow access to the fields of the data for that particular category.
The reason I've been thinking about having all five categories in the union type is essentially to be resilient to future changes, to make clear that the field could include data specific to any category. That might be overengineering it, though.

DylanSpOddball replied:
Per @adamodd's comment [here](https://github.com/CMSgov/easi-app/pull/2231#issuecomment-1767646782), I'm keeping the union type covering all five categories (which means keeping the placeholder fields in place, though I've changed them to be nullable, and specified that they'll always be null in https://github.com/CMSgov/easi-app/pull/2231/commits/cc22b92431c70cb24d34eb27c4b9e06cee93f2de) and keeping `categorySpecificData` as non-nullable on the `TRBAdminNote` type.

mynar7 reviewed (commented):
My preference for this admin note categorization feature would normally be to split out different resolvers for each note category (and probably tables/models as well in the BE) so that we could have useful type inference on both FE and BE.

If we took that approach, in theory, the FE could use the selected form category in a higher order component that wraps the admin note form to render different note form components that use the different note category mutations. This would hopefully prevent the FE form code from getting too complex while allowing for type safety and for GQL to handle the validation logic on both ends. (I'm not an expert by any means on the current implementation so I could be completely wrong here, admittedly.)

However, after speaking with Dylan, we're also planning on treating all admin notes as one data entity in the BE and returning them as a union type. In light of that, I think a good compromise would be to flatten the admin note input types so that all categorical fields are optional and then running validation in the BE as necessary. (The Figma also does not currently indicate any of the categorical data is required, though that may change in the future.)

I think the tradeoff is that, for better user experience, the FE may end up writing its own validation code for categorical data to match the BE instead of relying on the mutation call to throw a GQL error. Also, storing admin notes in a single table will probably introduce some type system gymnastics down the road in the backend from trying to enforce and associate specific fields with certain categories.

DylanSpOddball commented:
I paired with Lee today on this, notes:
* We're not sure on whether to use multiple mutations, one per category, or a single mutation that covers all categories. I've set up a meeting tomorrow (Tuesday) to talk about this and make a decision.
* We discussed having GraphQL types for each category, with the `TRBAdminNote` type being a union of all these types. This helps with the ergonomics of working with query results in frontend code a bit; the frontend code would need to check `adminNote.__typename` instead of `adminNote.categorySpecificData.__typename`. However, there's some issues with repetitive code in the GraphQL schema; I tried using a [GraphQL interface](https://graphql.org/learn/schema/#interfaces) to provide some type safety, but this creates problems with gqlgen's model binding, it can't bind a GraphQL interface to a Go struct (just a Go interface). [This branch](https://github.com/CMSgov/easi-app/tree/EASI-3360/trb-admin-note-union) has my experiments, but I don't think it's worth pursuing.
* (for the backend implementation) We'll go ahead with keeping a single Go struct and a single SQL table.

mynar7 commented:
> I paired with Lee today on this, notes:
> 
> * We're not sure on whether to use multiple mutations, one per category, or a single mutation that covers all categories. I've set up a meeting tomorrow (Tuesday) to talk about this and make a decision.

The benefit of the multiple mutations is that it provides type inference and validation through GQL that would be lost through a single mutation with multiple optional fields.

However, Dylan mentioned there could be some weirdness around switching between multiple mutations when trying to submit the different admin note types in the front-end. I think this could be alleviated by using different form components for each note type. [I made a demo to showcase this approach](https://codepen.io/mynar7/pen/yLGdxqO?editors=0010) ahead of the meeting today. @aterstriep @DylanSpOddball

DylanSpOddball commented:
Clay, Patrick, Ashley, Adam, Lee, and I met on 10/17 to decide on what approach to use with the createTRBAdminNote mutation - whether to use a single mutation with nullable fields vs. multiple mutations, one for each category. We decided to use multiple mutations - the frontend team has a few techniques they can use to make that manageable in the frontend code for EASI-3361, which @patrickseguraoddball is going to mention on that ticket.

I'll update this PR with multiple mutations for creating admin notes; @ClayBenson94 and I are going to talk with UX about whether the updateTRBAdminNote mutation is still needed.

DylanSpOddball commented:
> Clay, Patrick, Ashley, Adam, Lee, and I met on 10/17 to decide on what approach to use with the createTRBAdminNote mutation - whether to use a single mutation with nullable fields vs. multiple mutations, one for each category. We decided to use multiple mutations - the frontend team has a few techniques they can use to make that manageable in the frontend code for [EASI-3361](https://jiraent.cms.gov/browse/EASI-3361), which @patrickseguraoddball is going to mention on that ticket.
> 
> I'll update this PR with multiple mutations for creating admin notes; @ClayBenson94 and I are going to talk with UX about whether the updateTRBAdminNote mutation is still needed.

I've updated this PR with multiple new mutations and their associated input types, keeping the existing createTRBAdminNote mutation around to keep existing code functioning. I created the subtask EASI-3458 for cleaning up that mutation and its resolver later, once EASI-3361 and EASI-3662 are complete and it's no longer being used.

After Clay and I talked with UX, we decided to keep the updateTRBAdminNote mutation around, but not to update it for now; I've added comments outlining that.

Two outstanding questions:
1. Should the `TRBAdminNoteCategorySpecificData` union type include the types with no category-specific data, or exclude them? See [this comment thread](https://github.com/CMSgov/easi-app/pull/2231#discussion_r1360925848)
2. Does the nested `CreateTRBAdminNoteCommonFields` GraphQL type work with the frontend code that calls the mutations?

adamodd commented:
1. If they're the same amount of work to do then I think just include them.
2. If FE doesn't have a particular reason to have the `commonFields` property nested like that by the time of input, then I think it's a little more common to see those properties flattened into the main object

DylanSpOddball commented:
> 1. If they're the same amount of work to do then I think just include them.

Ok, I'll keep the existing code, then.

> 2. If FE doesn't have a particular reason to have the `commonFields` property nested like that by the time of input, then I think it's a little more common to see those properties flattened into the main object

Ok, did this in https://github.com/CMSgov/easi-app/pull/2231/commits/6f65b89e7e61908dcb7bf87edc7db4409a2a4678.

mynar7 commented on pkg/graph/schema.graphql:
```diff
@@ -2316,6 +2365,7 @@ type TRBAdminNote {
   noteText: HTML!
   author: UserInfo!
   isArchived: Boolean!
+  categorySpecificData: TRBAdminNoteCategorySpecificData!
```
I'm not 100% sure if it's possible, but I would lift this union up to be a union of the whole note type instead of just the categorical data. It adds repetition of common fields in each note, but we have 5 type definitions of categorical data anyway with this approach.

Put another way, if the front end has to check the category of each note regardless, why not flatten the objects and just check `note.__typename` instead?

If I had to pick between empty placeholders for categorical types and some repetition of fields in note data, I would prefer the latter. This allows for expanding general notes with new fields in the future without having to remove the placeholder, creates parity with the separate mutations, and creates sensible types for querying separate note types in the future if needed.

DylanSpOddball replied:
I experimented with this a bit; see [this comment](https://github.com/CMSgov/easi-app/pull/2231#issuecomment-1765408993) and [this branch](https://github.com/CMSgov/easi-app/tree/EASI-3360/trb-admin-note-union) (in particular, https://github.com/CMSgov/easi-app/commit/29735cfcff8727f2262550e3478db1704f914751). The problems with having `TRBAdminNote` be a union type are:
1. If the common fields are repeated between the different category-specific types, with a GraphQL interface to enforce that they all have the common fields, that interface breaks `gqlgen`'s model binding.
2. If all the common fields are repeated without any way to enforce that the category-specific types have the same fields, that opens up a lot of room for error if we have to change the common fields or add more in the future; we'd need to make sure we make the same changes across all five category-specific types. I can live with it on the input types because there's only two common fields (`trbRequestId` and `noteText`), but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency.
3. We'd have to create a separate new `TRBAdminNoteUnion` type as the union so existing queries don't break; the frontend code in EASI-3361 would then change to querying against that type, and then we'd need to clean up the old type later.

It might be possible to extract the common fields into a separate type, which each of the category-specific types would contain, so there's something like:
```gql
type TRBAdminNoteCommonFields {
  id: UUID!
  trbRequestId: UUID!
  category: TRBAdminNoteCategory!
  noteText: HTML!
  author: UserInfo!
  isArchived: Boolean!
  createdBy: String!
  createdAt: Time!
  modifiedBy: String
  modifiedAt: Time
}

type TRBAdminNoteGeneralRequest {
  commonFields: TRBAdminNoteCommonFields!
}

type TRBAdminNoteInitialRequestForm {
  commonFields: TRBAdminNoteCommonFields!
  appliesToBasicRequestDetails: Boolean!
  appliesToSubjectAreas: Boolean!
  appliesToAttendees: Boolean!
}

# types for the other three categories...

union TRBAdminNoteUnion = TRBAdminNoteGeneralRequest
  | TRBAdminNoteInitialRequestForm
  | TRBAdminNoteSupportingDocuments
  | TRBAdminNoteConsultSession
  | TRBAdminNoteAdviceLetter
```
But I'm not sure if having the nested fields (instead of flattening them out) would cause issues with the BE storage code and/or the FE form code.

In short: I think the current approach works well enough and doesn't cause any significant problems; other alternatives would either cause more problems or would require more work that probably wouldn't add much value even if they didn't cause issues.

mynar7 replied:
> but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency

Considering this is a return/query type for the FE, I don't see this causing any inconsistency issues in the data. In the worst case scenario, we forget to return a new field in one of the return types and have to add it. I think this could be solved with a comment noting common fields like you've already done. It would be nice to enforce with an interface, but I think the repetition is acceptable since it's a limitation of Apollo/GQL that we're running up against.

I think the more likely and problematic scenario is that we have a need to query by note category at some point in the future. That would mean either adding complexity to the current resolver and storage method to allow filtering, or making separate resolvers and storage methods and creating the new return types you're wanting to avoid creating anyway. Creating the separate resolvers and note types will avoid a category check on the FE, which leads me to believe that would be the preference if we needed that feature.

In other words, we can avoid the union type when querying notes of a specific category if we have separate note types, but if we put that union in a generic note type, the FE will always need to check the category when querying for any kind of note.

> We'd have to create a separate new TRBAdminNoteUnion type as the union so existing queries don't break; the frontend code in [EASI-3361](https://jiraent.cms.gov/browse/EASI-3361) would then change to querying against that type, and then we'd need to clean up the old type later.

This ticket isn't in progress, so I don't see the issue here.

DylanSpOddball replied:
> > but repeating the 9-10 common fields that are returned from queries seems like it introduces a lot of possibilities for inconsistency
> 
> Considering this is a return/query type for the FE, I don't see this causing any inconsistency issues in the data. In the worst case scenario, we forget to return a new field in one of the return types and have to add it. I think this could be solved with a comment noting common fields like you've already done. It would be nice to enforce with an interface, but I think the repetition is acceptable since it's a limitation of Apollo/GQL that we're running up against.

Inconsistency when making changes in the future is what I mean, because there wouldn't be any type safety or anything in the schema to keep those fields from getting out of sync, whether that's adding fields, removing fields, changing fields' type.

> I think the more likely and problematic scenario is that we have a need to query by note category at some point in the future. That would mean either adding complexity to the current resolver and storage method to allow filtering, or making separate resolvers and storage methods and creating the new return types you're wanting to avoid creating anyway. Creating the separate resolvers and note types will avoid a category check on the FE, which leads me to believe that would be the preference if we needed that feature.
> 
> In other words, we can avoid the union type when querying notes of a specific category if we have separate note types, but if we put that union in a generic note type, the FE will always need to check the category when querying for any kind of note.

For one thing, that's a pretty specific case of functionality we _might_ add in the future to worry about. For another, I'm not sure how having a separate type for each category would even help with that, unless we added top-level queries for each individual category? 

> > We'd have to create a separate new TRBAdminNoteUnion type as the union so existing queries don't break; the frontend code in [EASI-3361](https://jiraent.cms.gov/browse/EASI-3361) would then change to querying against that type, and then we'd need to clean up the old type later.
> 
> This ticket isn't in progress, so I don't see the issue here.

It's more work we'd have to do in the future. We'd also need to add an additional field to the `TRBRequest` type now, so the frontend ticket could switch to querying the category-specific data when they're ready.

DylanSpOddball replied:
My worry above about nested types being an issue isn't really applicable (since a `categorySpecificData` approach introduces nested types anyways), so we could have a type for the common fields. We'd still have the issues with needing to add a new type + new field on `TRBRequest` for the union (and cleaning it up later), though.

mynar7 replied:
> Inconsistency when making changes in the future is what I mean, because there wouldn't be any type safety or anything in the schema to keep those fields from getting out of sync, whether that's adding fields, removing fields, changing fields' type.

Yes, but those types would all sit next to each other in the schema file so I don't think keeping them in sync is that big of a concern.

> For one thing, that's a pretty specific case of functionality we might add in the future to worry about. 

For example, if users were viewing the advice letter, I could see a scenario where we're asked to add a feature to view the relevant notes for the advice letter.

> For another, I'm not sure how having a separate type for each category would even help with that, unless we added top-level queries for each individual category?

Top-level queries is what I am suggesting for each note category. This would return the individual note type and there's no need for the FE to validate category specific data. 

> It's more work we'd have to do in the future.

Yes, but I don't see it as a huge lift.

DylanSpOddball replied:
Lee, Clay, and I had a group chat about this - we're going with the current approach of having a union type just for the category-specific data.

mynar7 reviewed (commented):
This looks good overall, but I am curious if we could flatten the note types on the query side.

aterstriep reviewed (approved):
Changes look good. Thanks for all your work on this, @DylanSpOddball!
